### ISSUE TITLE
Enums are not created in new type alias syntax

### ISSUE THREAD

**Issue by Alc-Alc** [sqlalchemy/sqlalchemy#11305]
### Describe the bug

Using the new type alias syntax with a literal causes errors during the model creation. I would expect both the syntaxes to work OOTB as both are type aliases (check Additional Info for screenshot)

Tried adding `type_annotation_map` (code as seen in Additional Info) same error

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map

### SQLAlchemy Version in Use

2.0.29

### DBAPI (i.e. the database driver)

NA - Happens at model creation

### Database Vendor and Major Version

NA - Happens at model creation

### Python Version

3.12

### Operating system

OSX

### To Reproduce

```python
from typing import Literal, reveal_type
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

type Options484 = Literal["one", "two"]
Options = Literal["one", "two"]

reveal_type(Options)  # Runtime type is '_LiteralGenericAlias'
reveal_type(Options484)  # Runtime type is 'TypeAliasType'

class Base(DeclarativeBase): ...

class MyTable(Base):
    __tablename__ = "my_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    option_1: Mapped[Options]  # works
    option_2: Mapped[Options484]  # does not work
```

### Error

```
sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type typing.Literal['one', 'two'] inside the 'option_2' attribute Mapped annotation
``` 

### Additional context

Hovering over the aliases in reveal_type shows they are both picked up by Pylance (Pyright) as type aliases.

**For `Options`**

<img width="564" alt="image" src="https://github.com/sqlalchemy/sqlalchemy/assets/45509143/f58e7612-d8ea-47f5-8c2e-91115f71f177">

**For `Options484`**

<img width="568" alt="image" src="https://github.com/sqlalchemy/sqlalchemy/assets/45509143/b8b14b99-7a17-4f1e-8601-5406ee82fdeb">

Tried adding `type_annotation_map` for `TypeAliasType`, same error as shown in "Error".

```py
from typing import Literal, TypeAliasType
from enum import Enum
from sqlalchemy import Enum as SAEnum
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

type Options484 = Literal["one", "two"]
Options = Literal["one", "two"]

class Base(DeclarativeBase): ...


class MyTable(Base):
    __tablename__ = "my_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    option_1: Mapped[Options]  # works
    option_2: Mapped[Options484]  # does not work
    type_annotation_map = {
        TypeAliasType: SAEnum(Enum),
    }
```

PS: 484 should actually be 695, a reference to https://peps.python.org/pep-0695/ which describes the syntax. I got the PEPs confused. 🙃

**Comment by CaselIT** [comment]
Hi,

I fairly sure this has already been reported but I cannot find it.
Sadly python decided that for the typing feature it was the best thing to have a thousand different ways of doing anything, so it's quite difficult to support all the various incantations.

thanks for reporting

**Comment by CaselIT** [comment]
Also `TypeAliasType` is not a valid type here, since it's likely that everything defined with `type ...` becomes one. Indeed:
```py
>>> type x = int
>>> type(x)
<class 'typing.TypeAliasType'>
>>> type y = int | str | float
>>> type(x)
<class 'typing.TypeAliasType'>
```

**Comment by CaselIT** [comment]
It seems that one could use `__value__` to access what the type alias points to

**Comment by Alc-Alc** [comment]
> I fairly sure this has already been reported but I cannot find it.

I came across https://github.com/sqlalchemy/sqlalchemy/issues/10807 which felt similar but the issue was closed and it seems it was released in [2.0.25](https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-2.0.25)

> Added preliminary support for Python 3.12 pep-695 type alias structures, when resolving custom type maps for ORM Annotated Declarative mappings.

but I was still running into this so I opened a new one.

**Comment by CaselIT** [comment]
Ok, Thanks for finding that. Then I guess this is an unsupported case

**Comment by Alc-Alc** [comment]
@CaselIT I have made an attempt here at fixing this, if you feel this is a good starting point I am happy to raise a PR and apply any relevant suggestions. I added one new test (modified based on an existing test) and it passes.

https://github.com/sqlalchemy/sqlalchemy/compare/main...Alc-Alc:sqlalchemy:fix-11305?expand=1

**Comment by CaselIT** [comment]
That seems to work but I wonder if it could be easier to just do at the start of `_resolve_type`
```py
if is_pep695(python_type):
  return _resolve_type(python_type.__value__
```

This would handle anything is declared with the new type syntax.
It would also allow for alias of alias (not sure if value in this case is also a typealiastype or if it's resolved)

Feel free to open a PR. It's a good start (if you do can you also add a change log?)

**Comment by sqla-tester** [comment]
**Alc-Alc** has proposed a fix for this issue in the **main** branch:

**Enums are not created with pep695 literal** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5260

**Comment by zzzeek** [comment]
was there another issue or discussion where we went over this previously?

**Comment by sqla-tester** [comment]
**Alc-Alc** has proposed a fix for this issue in the **rel_2_0** branch:

**improve pep-695 inference including Enum support** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5278

**Comment by zzzeek** [comment]
OK mostly this just builds on 8772041cc6 /  #10807  which is very recent .  great thanks for the PR

**Comment by Alc-Alc** [comment]
> was there another issue or discussion where we went over this previously?

Yes, I noticed you figured it in the next comment. Just for completion sake, https://github.com/sqlalchemy/sqlalchemy/issues/11305#issuecomment-2074638338 states my reasoning behind opening this issue.

**Comment by CaselIT** [comment]
This seems like a wrong implementation.

Python is clearly telling us that `type ...` is something different from a simple assign
```py
In [16]: from typing import Literal
In [17]: Options = Literal["one", "two"]
In [18]: Options2 = Literal["one", "two"]
In [19]: Options == Options2
Out[19]: True
In [20]: type O = Literal["one", "two"]
In [21]: type O2 = Literal["one", "two"]
In [22]: O==O2
Out[22]: False
In [23]: O==Options
Out[23]: False
```

I think the spirit of `type ...` syntax is to behave like newtype.

**Comment by CaselIT** [comment]
So I think that the correct implementation is to require the type alias type to be inside the type map explicitly

**Comment by CaselIT** [comment]
The correct way of doing this should be
```py
import enum
from typing import Literal, reveal_type
from sqlalchemy import Enum
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

type Options484 = Literal["one", "two"]
Options = Literal["one", "two"]

reveal_type(Options)  # Runtime type is '_LiteralGenericAlias'
reveal_type(Options484)  # Runtime type is 'TypeAliasType'

class Base(DeclarativeBase):
    type_annotation_map = {Options484: Enum(enum.Enum)}

class MyTable(Base):
    __tablename__ = "my_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    option_1: Mapped[Options]  # works
    option_2: Mapped[Options484]  # does not work
```

**Comment by CaselIT** [comment]
This is how #11955 will change the behaviour of it

**Comment by sqla-tester** [comment]
**Frazer McLean** has proposed a fix for this issue in the **main** branch:

**General improvement on annotated declarative** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5513

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**General improvement on annotated declarative** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5601

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**General improvement on annotated declarative** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5602

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Remove fallbacks from the previous typing change** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5603

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**edits and reorganization for union/pep695 typing docs** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5641

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**edits and reorganization for union/pep695 typing docs** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5642
